### PR TITLE
8267186: Add string deduplication support to ZGC

### PR DESCRIPTION
**Pull request by pliden**
This change adds support for string deduplication to ZGC. It's a pretty straight forward change, but to make reviewing even easier it is broken up into two commits:

__ZGC: Introduce ZMarkContext__
This commit just moves the `ZMarkCache` into the new `ZMarkContext` class, and we now pass a `ZMarkContext*` around instead of a `ZMarkCache*`. The `ZMarkContext` class is a more general container for worker-local data, which in the next commit will be extended to also include the `StringDedup::Requests` queue.

__8267186: Add string deduplication support to ZGC__
This commits adds the actual string dedup functionality and enables relevant tests. We use the `deduplication_requested` bit in the `String` object to filter out `Strings` we've already attempted to deduplicate.

Testing:
- Passes all string dedup tests.
- Passes Tier1-7 with ZGC on Linux/x86_64 (with -XX:+UseStringDeduplication enabled by default to get better exposure).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8267186](https://bugs.openjdk.java.net/browse/JDK-8267186): Add string deduplication support to ZGC

### Reviewers
 * [Erik Österlund](https://openjdk.java.net/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to c8cd63ee9c2e2e9d9feeb44e86b022d4bd4f4d31
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to c8cd63ee9c2e2e9d9feeb44e86b022d4bd4f4d31
 * [Stefan Karlsson](https://openjdk.java.net/census#stefank) (@stefank - **Reviewer**) ⚠️ Review applies to c164f2c973f148e489634cca1b8767c3c8aca018

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5029/head:pull/5029` \
`$ git checkout pull/5029`

Update a local copy of the PR: \
`$ git checkout pull/5029` \
`$ git pull https://git.openjdk.java.net/jdk pull/5029/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5029`

View PR using the GUI difftool: \
`$ git pr show -t 5029`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5029.diff">https://git.openjdk.java.net/jdk/pull/5029.diff</a>

</details>

### CODE REVIEW

**Review by fisk — CHANGES_REQUESTED** [review]
Now that we have a context object for marking, it would be nice if it also contained the stripes and stacks that we pass around everywhere, that we originally determine just the lines below where the context object is created.

**Comment by pliden** [comment]
> Now that we have a context object for marking, it would be nice if it also contained the stripes and stacks that we pass around everywhere, that we originally determine just the lines below where the context object is created.

Good suggestion, will fix that.

**Review by fisk — APPROVED** [review]
Looks good.

File: src/hotspot/share/gc/z/zMarkContext.inline.hpp [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#ifndef SHARE_GC_Z_ZMARKCONTEXT_INLINE_HPP
+#define SHARE_GC_Z_ZMARKCONTEXT_INLINE_HPP
+
+#include "classfile/javaClasses.inline.hpp"
+#include "gc/z/zMarkCache.inline.hpp"
+#include "gc/z/zMarkContext.hpp"
```
Comment (kimbarrett) [inline src/hotspot/share/gc/z/zMarkContext.inline.hpp:29]: The associated .hpp file is supposed to be included first by a .inline.hpp file, per JDK-8267464.

**Review by kimbarrett — APPROVED** [review]
Looks good, except for the include order in zMarkContext.inline.hpp.

File: src/hotspot/share/gc/z/zMarkContext.inline.hpp (hunk 1 above)
Comment (pliden) [reply src/hotspot/share/gc/z/zMarkContext.inline.hpp]: Ah, forgot about that new rule. Fixed.

**Review by stefank — APPROVED** [review]
Looks good.

ZMarkContext serves multiple purposes in the marking code, and if we (maybe in the future) add code for those purposes then it risk being a class of multiple responsibilities. Therefore, I think it would be good to let this class be a simpler data carrier, and move the extra code/logic in ZMarkContext::try_deduplicate out to ZMark, or maybe even a new ZStringDedup class.

**Comment by pliden** [comment]
> [...] I think it would be good to let this class be a simpler data carrier, and move the extra code/logic in ZMarkContext::try_deduplicate out to ZMark, or maybe even a new ZStringDedup class.

Fixed.

**Comment by pliden** [comment]
Thanks @fisk, @kimbarrett and @stefank for reviewing!

/integrate

**Pull request closed without merging**